client/src/js-sdk/sdk.js

```diff
@@ -1,49 +1,271 @@
 import db, { FIREBASE_FUNCTION_BASE_URL } from './fire.js';
+import firebase from 'firebase'
 import axios from 'axios';
+import uuidv4 from 'uuid/v4';
 
 // Listeners
-export function listenToZone(player, zone, callback) {
-  db.doc(`Games/game1/Players/${player}/Zones/${zone}`)
+export function listenToZone(gameId, player, zone, callback) {
+  db.doc(`Games/${gameId}/Players/${player}/Zones/${zone}`)
     .onSnapshot((doc) => {
       callback(doc.data())
     });
 }
 
-export function listenToPlayer(player, callback) {
-  db.doc(`Games/game1/Players/${player}`)
+export function listenToPlayer(gameId, player, callback) {
+  db.doc(`Games/${gameId}/Players/${player}`)
     .onSnapshot((doc) => {
       callback(doc.data())
     });
 }
 
 // Game functions
-export function tap(card) {
+// Card functions
+export function moveCardToZone(gameId, card, targetZone) {
+  if (card["state.zone"] === targetZone) {
+    return;
+  }
+
+  remove(gameId, card); // remove from current zone before moving
+  var playerName = card["state.owner"];
+  card["state.zone"] = targetZone
+  db.doc(`Games/${gameId}/Players/${playerName}/Zones/${targetZone}`)
+    .set({
+      [card.id]: card
+    }, { merge: true })
+}
+
+function setSingleCardPosition(gameId, card, newPosition) {
+  var playerName = card["state.owner"];
+  var zoneName = card["state.zone"];
+  card["state.position"] = newPosition
+  db.doc(`Games/${gameId}/Players/${playerName}/Zones/${zoneName}`)
+    .update({
+      [card.id]: card
+    })
+}
+
+export function setCardPosition(gameId, card, newPosition) {
```

Comment:
I think it's rare for a card to be placed in a specific position (I can only recall [some cards](https://scryfall.com/search?as=grid&order=name&q=oracle%3A%22third+from+the+top%22) that instructs you to place a card 3rd from the top). Usually it's place a card at the top or bottom (if there are multiple cards that go to the top or bottom, e.g. as part of [Gishath's ability](https://scryfall.com/card/xln/222/gishath-suns-avatar), then the action could be composed by multiple "putAtTop/Bottom" function calls), in which case, only the card being placed needs to update position (recall that the position value is relative, and is not bound to 0-99). Going back to "third from the top", if we default our positions for the library to be spaced out 10 (this is arbitrary, it could be 100 or 1000) apart, e.g. top card is 0, second card is 10, etc, then we could insert third from the top by setting the position to 15. Whenever we shuffle the library (or we run out of spacing for just inserting), then we could cleanup the position values.

Reply:
I think it's easier to just do this than have to go through and check for when we will have to cleanup the values. Otherwise we'd have to go through and check that the value we want isn't taken, then check again, etc. until we find that there is no space.

Reply:
My original thinking was that since the data is cached, it would be easy to find the mid point between two taken position values, but the SDK does not do any caching itself, so it wouldn't work unless that changed.